contracts/SimpleQuests.sol

```diff
@@ -51,41 +54,64 @@ contract SimpleQuests is Initializable, AccessControlUpgradeable {
     uint8 public constant VAR_REPUTATION_LEVEL_4 = 22;
     uint8 public constant VAR_REPUTATION_LEVEL_5 = 23;
     uint8 public constant VAR_SKIP_QUEST_STAMINA_COST = 30;
+    uint8 public constant VAR_WEEKLY_COMPLETIONS_GOAL = 31;
     uint256 internal constant NFTVAR_SIMPLEQUEST_PROGRESS = 101;
     uint256 internal constant NFTVAR_SIMPLEQUEST_TYPE = 102;
     uint256 internal constant NFTVAR_REPUTATION = 103;
 
     struct Quest {
         uint256 id;
         Rarity tier;
-        RequirementType requirementType;
-        Rarity requirementRarity;
+        ItemType requirementType;
+        uint256 requirementRarity;
         uint256 requirementAmount;
-        RewardType rewardType;
-        Rarity rewardRarity;
+        address requirementExternalAddress;
+        ItemType rewardType;
+        uint256 rewardRarity;
         uint256 rewardAmount;
+        address rewardExternalAddress;
         uint256 reputationAmount;
     }
 
-    enum RequirementType{NONE, WEAPON, JUNK, DUST, TRINKET, SHIELD, STAMINA, SOUL, RAID}
-    enum RewardType{NONE, WEAPON, JUNK, DUST, TRINKET, SHIELD, EXPERIENCE, SOUL}
+    struct Reward {
+        uint256 id;
+        ItemType rewardType;
+        uint256 rewardRarity;
+        uint256 rewardAmount;
+        address rewardExternalAddress;
+        uint256 reputationAmount;
+    }
+
+    enum ItemType{NONE, WEAPON, JUNK, DUST, TRINKET, SHIELD, STAMINA, SOUL, RAID, EXPERIENCE, EXTERNAL, EXTERNAL_HOLD}
     enum Rarity{COMMON, UNCOMMON, RARE, EPIC, LEGENDARY}
 
     uint256 public nextQuestID;
+    uint256 public nextRewardID;
 
     mapping(uint256 => uint256[]) public questTemplates;
     mapping(uint256 => Quest) public quests;
+    mapping(uint256 => uint256) public questSupplies;
+    mapping(uint256 => uint256) public questDeadlines;
+    mapping(uint256 => uint256) public questIndexes;
     mapping(uint256 => uint256) public characterQuest;
     mapping(uint256 => uint256[4]) public tierChances;
     mapping(uint256 => uint256) public vars;
     mapping(uint256 => uint256) public lastFreeSkipUsage;
+    mapping(address => mapping(uint256 => uint256)) public weeklyCompletions; //user to week to completions
+    mapping(address => mapping(uint256 => bool)) public weeklyRewardClaimed;
+    mapping(uint256 => Reward) public rewards;
```

Comment:
we could map questitems to the same key as quest index to store input and output without nesting structs.
As for the weekly rewards, we probably want to pick one from an array that stores various possible rewards. Similar to how we pick a roll within the questtemplate tier arrays.

Reply:
Again, the issue with questItem struct, as I mentioned earlier is with Stack too deep in some methods that use too much local variables to return quest to frontend. I do understand your logic with indexes